File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/BottomNavigationViewItemSelectionsObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.BottomNavigationView.OnNavigationItemSelectedListener;
import android.view.Menu;
import android.view.MenuItem;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class BottomNavigationViewItemSelectionsObservable extends Observable<MenuItem> {
  private final BottomNavigationView view;

  BottomNavigationViewItemSelectionsObservable(BottomNavigationView view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super MenuItem> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnNavigationItemSelectedListener(listener);

    // Emit initial item, if one can be found
    Menu menu = view.getMenu();
    for (int i = 0, count = menu.size(); i < count; i++) {
      MenuItem item = menu.getItem(i);
      if (item.isChecked()) {
        observer.onNext(item);
        break;
      }
    }
  }

  static final class Listener extends MainThreadDisposable
      implements OnNavigationItemSelectedListener {
    private final BottomNavigationView bottomNavigationView;
    private final Observer<? super MenuItem> observer;

    Listener(BottomNavigationView bottomNavigationView, Observer<? super MenuItem> observer) {
      this.bottomNavigationView = bottomNavigationView;
      this.observer = observer;
    }

    @Override public boolean onNavigationItemSelected(@NonNull MenuItem item) {
      if (!isDisposed()) {
        observer.onNext(item);
      }
      return true;
    }

    @Override protected void onDispose() {
      bottomNavigationView.setOnNavigationItemSelectedListener(null);
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/BottomNavigationViewItemSelectionsObservable.kt
@file:JvmName("RxBottomNavigationView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import android.view.MenuItem
import androidx.annotation.CheckResult
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.bottomnavigation.BottomNavigationView.OnNavigationItemSelectedListener
import com.jakewharton.rxbinding3.internal.checkMainThread
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

/**
 * Create an observable which emits the selected item in `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* If an item is already selected, it will be emitted immediately on subscribe.
 */
@CheckResult
fun BottomNavigationView.itemSelections(): Observable<MenuItem> {
  return BottomNavigationViewItemSelectionsObservable(this)
}

private class BottomNavigationViewItemSelectionsObservable(
  private val view: BottomNavigationView
) : Observable<MenuItem>() {

  override fun subscribeActual(observer: Observer<in MenuItem>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setOnNavigationItemSelectedListener(listener)

    // Emit initial item, if one can be found
    val menu = view.menu
    for (i in 0 until menu.size()) {
      val item = menu.getItem(i)
      if (item.isChecked) {
        observer.onNext(item)
        break
      }
    }
  }

  private class Listener(
    private val bottomNavigationView: BottomNavigationView,
    private val observer: Observer<in MenuItem>
  ) : MainThreadDisposable(), OnNavigationItemSelectedListener {

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
      if (!isDisposed) {
        observer.onNext(item)
      }
      return true
    }

    override fun onDispose() {
      bottomNavigationView.setOnNavigationItemSelectedListener(null)
    }
  }
}
